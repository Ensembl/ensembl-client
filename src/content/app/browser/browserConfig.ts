/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ReactComponent as zoomInIcon } from 'static/img/browser/zoom-in.svg';
import { ReactComponent as zoomOutIcon } from 'static/img/browser/zoom-out.svg';
import { ReactComponent as navigateUpIcon } from 'static/img/browser/navigate-up.svg';
import { ReactComponent as navigateDownIcon } from 'static/img/browser/navigate-down.svg';
import { ReactComponent as navigateLeftIcon } from 'static/img/browser/navigate-left.svg';
import { ReactComponent as navigateRightIcon } from 'static/img/browser/navigate-right.svg';
import { ReactComponent as trackHeightIcon } from 'static/img/browser/icon_tracks_height.svg';
import { ReactComponent as trackLockIcon } from 'static/img/browser/icon_tracks_lock_open.svg';
import { ReactComponent as trackHighlightIcon } from 'static/img/browser/icon_tracks_highlight.svg';
import { ReactComponent as trackMoveIcon } from 'static/img/browser/icon_tracks_move.svg';

import { BrowserNavAction } from 'src/content/app/browser/browserState';

export type BrowserNavItem = {
  description: string;
  detail: {
    [key: string]: number;
  };
  icon: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  name: BrowserNavAction;
};

export const browserNavConfig: BrowserNavItem[] = [
  {
    description: 'navigate up',
    detail: {
      move_up_px: 50
    },
    icon: navigateUpIcon,
    name: BrowserNavAction.MOVE_UP
  },
  {
    description: 'navigate down',
    detail: {
      move_down_px: 50
    },
    icon: navigateDownIcon,
    name: BrowserNavAction.MOVE_DOWN
  },
  {
    description: 'zoom out',
    detail: {
      zoom_by: -0.3
    },
    icon: zoomOutIcon,
    name: BrowserNavAction.ZOOM_OUT
  },
  {
    description: 'zoom in',
    detail: {
      zoom_by: 0.3
    },
    icon: zoomInIcon,
    name: BrowserNavAction.ZOOM_IN
  },
  {
    description: 'navigate left',
    detail: {
      move_left_px: 50
    },
    icon: navigateLeftIcon,
    name: BrowserNavAction.MOVE_LEFT
  },
  {
    description: 'navigate right',
    detail: {
      move_right_px: 50
    },
    icon: navigateRightIcon,
    name: BrowserNavAction.MOVE_RIGHT
  }
];

export const browserTrackConfig = {
  trackHeightIcon: {
    description: 'Change track height',
    icon: trackHeightIcon
  },
  trackLockIcon: {
    description: 'Lock track',
    icon: trackLockIcon
  },
  trackHighlightIcon: {
    description: 'Highlight track',
    icon: trackHighlightIcon
  },
  trackMoveIcon: {
    description: 'Move track',
    icon: trackMoveIcon
  }
};
