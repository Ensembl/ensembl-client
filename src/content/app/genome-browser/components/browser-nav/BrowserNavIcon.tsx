/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { memo } from 'react';
import { OutgoingAction } from '@ensembl/ensembl-genome-browser';
import classNames from 'classnames';

import useGenomeBrowser from 'src/content/app/genome-browser/hooks/useGenomeBrowser';

import ImageButton from 'src/shared/components/image-button/ImageButton';

import { browserNavIconActionMap } from 'src/content/app/genome-browser/state/browser-nav/browserNavSlice';

import { BrowserNavItem } from 'src/content/app/genome-browser/components/browser-nav/browserNavConfig';
import { Status } from 'src/shared/types/status';

import iconStyles from './BrowserNavIcon.scss';

type BrowserNavIconProps = {
  browserNavItem: BrowserNavItem;
  enabled: boolean;
  className?: string;
};

export const BrowserNavIcon = (props: BrowserNavIconProps) => {
  const { genomeBrowser } = useGenomeBrowser();

  if (!genomeBrowser) {
    return null;
  }

  const { browserNavItem, enabled } = props;
  const { icon } = browserNavItem;

  const action = {
    type: browserNavIconActionMap[browserNavItem.name],
    payload: browserNavItem.detail
  } as OutgoingAction;

  const navigateBrowser = () => {
    if (enabled) {
      genomeBrowser.send(action);
    }
  };

  const iconStatus = enabled ? Status.DEFAULT : Status.DISABLED;

  return (
    <div className={classNames(iconStyles.browserNavIcon, props.className)}>
      <ImageButton
        status={iconStatus}
        description={browserNavItem.description}
        onClick={navigateBrowser}
        image={icon}
      />
    </div>
  );
};

export default memo(BrowserNavIcon);
