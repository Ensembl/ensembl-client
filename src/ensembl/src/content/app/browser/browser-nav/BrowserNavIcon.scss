.browserNavIcon {
  display: inline-block;
  margin-left: 10px;

  button {
    & > img {
      height: 25px;
      width: 25px;
    }
  }
}
