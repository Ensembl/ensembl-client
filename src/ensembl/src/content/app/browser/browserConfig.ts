import { ReactComponent as zoomInIcon } from 'static/img/browser/zoom-in.svg';
import { ReactComponent as zoomOutIcon } from 'static/img/browser/zoom-out.svg';
import { ReactComponent as navigateUpIcon } from 'static/img/browser/navigate-up.svg';
import { ReactComponent as navigateDownIcon } from 'static/img/browser/navigate-down.svg';
import { ReactComponent as navigateLeftIcon } from 'static/img/browser/navigate-left.svg';
import { ReactComponent as navigateRightIcon } from 'static/img/browser/navigate-right.svg';
import { ReactComponent as trackHeightBtn } from 'static/img/browser/icon_tracks_height.svg';
import { ReactComponent as trackLockBtn } from 'static/img/browser/icon_tracks_lock_open.svg';
import { ReactComponent as trackHighlightBtn } from 'static/img/browser/icon_tracks_highlight.svg';
import { ReactComponent as trackMoveBtn } from 'static/img/browser/icon_tracks_move.svg';

export type BrowserNavItem = {
  description: string;
  detail: {
    [key: string]: number;
  };
  icon: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  name: string;
};

export const browserNavConfig: BrowserNavItem[] = [
  {
    description: 'navigate up',
    detail: {
      move_up_px: 50
    },
    icon: navigateUpIcon,
    name: 'navigate-up'
  },
  {
    description: 'navigate down',
    detail: {
      move_down_px: 50
    },
    icon: navigateDownIcon,
    name: 'navigate-down'
  },
  {
    description: 'zoom out',
    detail: {
      zoom_by: -0.3
    },
    icon: zoomOutIcon,
    name: 'zoom-out'
  },
  {
    description: 'zoom in',
    detail: {
      zoom_by: 0.3
    },
    icon: zoomInIcon,
    name: 'zoom-in'
  },
  {
    description: 'navigate left',
    detail: {
      move_left_px: 50
    },
    icon: navigateLeftIcon,
    name: 'navigate-left'
  },
  {
    description: 'navigate right',
    detail: {
      move_right_px: 50
    },
    icon: navigateRightIcon,
    name: 'navigate-right'
  }
];

export const browserTrackConfig = {
  trackHeightBtn: {
    description: 'Change track height',
    icon: trackHeightBtn
  },
  trackLockBtn: {
    description: 'Lock track',
    icon: trackLockBtn
  },
  trackHighlightBtn: {
    description: 'Highlight track',
    icon: trackHighlightBtn
  },
  trackMoveBtn: {
    description: 'Move track',
    icon: trackMoveBtn
  }
};
