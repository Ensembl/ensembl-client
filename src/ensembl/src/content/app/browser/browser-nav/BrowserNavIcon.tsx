import React, { FunctionComponent, memo } from 'react';

import browserMessagingService from 'src/content/app/browser/browser-messaging-service';
import { BrowserNavItem } from '../browserConfig';

import iconStyles from './BrowserNavIcon.scss';

type BrowserNavIconProps = {
  browserNavItem: BrowserNavItem;
  enabled: boolean;
};

export const BrowserNavIcon: FunctionComponent<BrowserNavIconProps> = (
  props: BrowserNavIconProps
) => {
  const { browserNavItem, enabled } = props;
  const { detail, icon } = browserNavItem;

  const navigateBrowser = () => {
    if (enabled) {
      browserMessagingService.send('bpane', detail);
    }
  };

  const iconUrl = enabled ? icon.on : icon.off;

  return (
    <button
      className={iconStyles.browserNavIcon}
      title={browserNavItem.description}
      onClick={navigateBrowser}
    >
      <img src={iconUrl} alt={browserNavItem.description} />
    </button>
  );
};

export default memo(BrowserNavIcon);
