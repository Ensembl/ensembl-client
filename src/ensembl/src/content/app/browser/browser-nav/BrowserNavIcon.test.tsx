/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { BrowserNavIcon } from './BrowserNavIcon';
import browserMessagingService from 'src/content/app/browser/browser-messaging-service';
import { browserNavConfig } from '../browserConfig';

describe('<BrowserNavIcon />', () => {
  const browserNavItem = browserNavConfig[0];

  test('sends navigation message when clicked', () => {
    jest.spyOn(browserMessagingService, 'send');

    const { container } = render(
      <BrowserNavIcon browserNavItem={browserNavItem} enabled={true} />
    );
    const button = container.querySelector('button') as HTMLButtonElement;

    userEvent.click(button);
    expect(browserMessagingService.send).toHaveBeenCalledTimes(1);
    expect(browserMessagingService.send).toHaveBeenCalledWith(
      'bpane',
      browserNavItem.detail
    );

    (browserMessagingService.send as any).mockRestore();
  });
});
