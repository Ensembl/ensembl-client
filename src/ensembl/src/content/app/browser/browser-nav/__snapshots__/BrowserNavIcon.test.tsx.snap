// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowserNavIcon /> renders correctly 1`] = `
<dd
  className="browserNavIcon"
>
  <button
    onClick={[Function]}
    title="navigate up"
  >
    <img
      alt="navigate up"
      src={Object {}}
    />
  </button>
</dd>
`;
